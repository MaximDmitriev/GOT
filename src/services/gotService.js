export default class gotService {
    constructor() {
        this._apiURL = "https://www.anapioficeandfire.com/api";
    }

    async getResourse(url) {
        const res = await fetch(this._apiURL + url);

        if(!res.ok) {
            throw new Error(res.status);  
        }

        return await res.json();
    }

    async getAllCharacters() {
        const res = await this.getResourse("/characters?page=7&pageSize=10");
        return res.map(this._transformCharacter);
    }

    async getCharacter(id) {
        const char = await this.getResourse(`/characters/${id}`);
        return this._transformCharacter(char);
    }

    async getAllBooks() {
        const res = await this.getResourse("/characters");
        return res.map(this._transformBook);
    }

    async getBook(id) {
        const book = await this.getResourse(`/characters/${id}`);
        return this._transformBook(book);
    }

    async getAllHouses() {
        const res = await this.getResourse("/characters");
        return res.map(this._transformHouse);
    }

    async getHouse(id) {
        const house = await this.getResourse(`/characters/${id}`);
        return this._transformHouse(house);
    }

    _transformCharacter(char) {
        return {
            name: char.name === "" ? "no data" : char.name,
            gender: char.gender === "" ? "no data" : char.gender,
            born: char.born === "" ? "no data" : char.born,
            died: char.died === "" ? "no data" : char.died,
            culture: char.culture === "" ? "no data" : char.culture,
            id: char.url.slice(-2)
        }
    }

    _transformHouse(house) {
        return {
            name: house.name,
            region: house.region,
            words: house.words,
            titles: house.titles,
            overlords: house.overlords,
            ancestralWeapons: house.ancestralWeapons
        }
    }

    _transformBook(book) {
        return {
            name: book.name,
            numberOfPages: book.numberOfPages,
            publiser: book.publiser,
            released: book.released
        }
    }
}

