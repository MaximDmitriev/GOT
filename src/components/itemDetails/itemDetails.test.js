import React from 'react';
import ItemDetails from "./itemDetails";
import {shallow} from "enzyme";
import gotService from "../../services/gotService";

describe("Testing <ItemDetails />", () => {
    const items = shallow(<ItemDetails item="book" />);
    const getData = new gotService();

    describe("Testing Shapshots & state",() => {

        it("ItemDetails have rendered", () => {
            expect(items).toMatchSnapshot();
        });
    
        it("ItemDetails state 'item' is empty", () => {
            expect(items.state().item).toBeNil();
        });
    
        it("ItemDetails state 'loading' is false", () => {
            expect(items.state().loading).toBeFalse();
        });
    
        it("ItemDetails state 'error' is false", () => {
            expect(items.state().error).toBeFalse();
        });

        it("ItemDetails props 'item' is string", () => {
            expect(items.instance().props.item).toBeString();
        });
    });

    describe("Handlers tests", () => {
        it("Testing itemUpdate", () => {
            items.instance().itemUpdate();
            expect(items.state().loading).toBeFalse();
        });
        
        it("Testing itemUpdate async", () => {
            const id = 4;
            items.instance().itemUpdate(getData.getAllBooks(id)
            .then(res => expect(items.state().item).toBe(res)));     
        });

        it("Testing componentDidCatch", () => {
            items.instance().componentDidCatch();
            expect(items.state().error).toBeTrue();
        });
    });

});
