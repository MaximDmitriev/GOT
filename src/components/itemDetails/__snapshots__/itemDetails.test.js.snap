// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <ItemDetails /> Testing Shapshots & state ItemDetails have rendered 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ItemDetails
    item="book"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        "Choose ",
        undefined,
        " from list",
      ],
      "className": "rounded",
    },
    "ref": null,
    "rendered": Array [
      "Choose ",
      undefined,
      " from list",
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-htpNat",
        "isStatic": true,
        "rules": Array [
          "
    font-size: 26px;
    background-color: #fff;
    padding: 0 20px;
",
        ],
      },
      "displayName": "styled.span",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-htpNat",
      "target": "span",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          "Choose ",
          undefined,
          " from list",
        ],
        "className": "rounded",
      },
      "ref": null,
      "rendered": Array [
        "Choose ",
        undefined,
        " from list",
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-htpNat",
          "isStatic": true,
          "rules": Array [
            "
    font-size: 26px;
    background-color: #fff;
    padding: 0 20px;
",
          ],
        },
        "displayName": "styled.span",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-htpNat",
        "target": "span",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
